File: modules/desktop/home/hyprland/eww/scripts/eww_bar_backend/inhibitors.py
import subprocess


HYPRIDLE_INHIBIT_SERVICE = "eww-hypridle-inhibit.service"
LID_INHIBIT_SERVICE = "eww-lid-inhibit.service"


def bool_state(value):
    return "true" if value else "false"


def service_active(service):
    result = subprocess.run(
        ["systemctl", "--user", "is-active", "--quiet", service],
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        check=False,
    )
    return result.returncode == 0


def set_service_active(service, enabled):
    action = "start" if enabled else "stop"
    result = subprocess.run(
        ["systemctl", "--user", action, service],
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        check=False,
    )
    if result.returncode != 0:
        raise RuntimeError(f"systemctl --user {action} {service} failed")
    return service_active(service)


def idle_inhibited_state():
    return bool_state(service_active(HYPRIDLE_INHIBIT_SERVICE))


def lid_inhibited_state():
    return bool_state(service_active(LID_INHIBIT_SERVICE))


def set_idle_inhibited(enabled):
    return bool_state(set_service_active(HYPRIDLE_INHIBIT_SERVICE, enabled))


def set_lid_inhibited(enabled):
    return bool_state(set_service_active(LID_INHIBIT_SERVICE, enabled))


def toggle_idle_inhibited():
    return set_idle_inhibited(not service_active(HYPRIDLE_INHIBIT_SERVICE))

File: modules/desktop/home/hyprland/eww/go/internal/collect/inhibitors.go
package collect

import (
	"fmt"
	"time"
)

// The two systemd user units that hold the idle and lid inhibitors. Starting
// and stopping a unit IS the state: nothing is cached, because the user can
// change it from outside the bar.

const (
	hypridleInhibitService = "eww-hypridle-inhibit.service"
	lidInhibitService      = "eww-lid-inhibit.service"
)

// systemctlTimeout exists because a hung systemctl would otherwise block a
// control-socket worker forever, and the socket has a bounded pool.
const systemctlTimeout = 5 * time.Second

// BoolState renders a flag the way eww reads it, as a string.
func BoolState(value bool) string {
	if value {
		return "true"
	}
	return "false"
}

func ServiceActive(service string) bool {
	return RunStatus(systemctlTimeout, "systemctl", "--user", "is-active", "--quiet", service) == 0
}

// SetServiceActive returns the state it reads back rather than the state it asked
// for: `systemctl start` on a unit whose ExecStart fails still exits zero for a
// Type=oneshot with RemainAfterExit, so the only honest answer is to ask again.
func SetServiceActive(service string, enabled bool) (bool, error) {
	action := "stop"
	if enabled {
		action = "start"
	}
	if code := RunStatus(systemctlTimeout, "systemctl", "--user", action, service); code != 0 {
		return false, fmt.Errorf("systemctl --user %s %s failed", action, service)
	}
	return ServiceActive(service), nil
}

func IdleInhibitedState() string { return BoolState(ServiceActive(hypridleInhibitService)) }

func LidInhibitedState() string { return BoolState(ServiceActive(lidInhibitService)) }

func SetIdleInhibited(enabled bool) (string, error) {
	active, err := SetServiceActive(hypridleInhibitService, enabled)
	if err != nil {
		return "", err
	}
	return BoolState(active), nil
}

func SetLidInhibited(enabled bool) (string, error) {
	active, err := SetServiceActive(lidInhibitService, enabled)
	if err != nil {
		return "", err
	}
	return BoolState(active), nil
}

func ToggleIdleInhibited() (string, error) {
	return SetIdleInhibited(!ServiceActive(hypridleInhibitService))
}
